[docs] Base Portal style revisions and final review (#32157)
mui/material-ui: docs/data/base/components/portal/portal.md

# Portal

<p class="description">The <code>Portal</code> component renders its children into a new subtree outside of the current DOM hierarchy.</p>

> The children of the `Portal` component will be appended to the specified container.

Author: Why would a developer need/want to use this component? I think it's worth explaining here.

Reviewer: Maybe it's worth pointing out that our Portal is just a component version of [`React.createPortal`](https://reactjs.org/docs/portals.html) and was created for convenience. All the use cases of React's portal apply here as well.

Author: That is very helpful information @michaldudak! I'll draft up a new version to try to explain this.

# Portal

<p class="description">The <code>Portal</code> component lets you render its children into a DOM node that exists outside of its own DOM hierarchy.</p>

`Portal` is a utility component built around [React's `createPortal()` API](https://reactjs.org/docs/portals.html).
It gives you the functionality of `createPortal()` in a convenient component form.

> 💡 According to [the React docs](https://reactjs.org/docs/portals.html), portals are useful when "you need the child element to visually 'break out' of its container."

The `Portal` component is used internally by Material UI's [`Modal`](/components/modal/) and [`Popper`](/components/popper/) components.

## Basic usage

> Normally, child elements are appended to the nearest parent node.

Reviewer: IMO something along these lines would read better:

```suggestion
Normally, children of a component are rendered within this component's DOM tree.
```

This is one of the places where having a diagram could make the concept much easier to understand.

But sometimes it's necessary to mount a child to a different location in the DOM.

> The `Portal` component accepts a `container` prop that passes a `ref` to the DOM node where its children will be mounted.

Author: I hope I explained this correctly! It's a lot of jargon to pin down. 😅

Reviewer 2: That could just be: 
```suggestion
Portal accepts a `container` prop that passes a `ref` to the DOM node where its children will be mounted.
```
(We've already established that Portal is a React component.)

Also is that the right "direction"? Isn't it the ref to the dom node where the children will be mounted passed to the Portal with this prop?

# Portal

> <p class="description">The <code>Portal</code> component lets you render its children into a DOM node that exists outside of its own DOM hierarchy.</p>

Reviewer 3: ```suggestion
<p class="description">The portal component lets you render its children into a DOM node that exists outside of its own DOM hierarchy.</p>
```

This doesn't work with Google:

```html
<meta name="description" content="The &lt;code&gt;Portal&lt;/code&gt; component lets you render its children into a DOM node that exists outside of its own DOM hierarchy."/>
```

Reviewer 3: It's possible to duplicate the description, having one specific for SEO, but personally, I would encourage we thrown when `<code>` or any HTML element is used in the description.

Author: Sounds good. Two descriptions could get messy pretty quickly. I think we should still capitalize it here at least, since we're talking about our specific component rather than the general concept.

> `Portal` is a utility component built around [React's `createPortal()` API](https://reactjs.org/docs/portals.html).

Reviewer 3: It think that we can keep the content more "lightweight" (easier to read, less style change interruption) with:

```suggestion
Portal is a utility component built around [React's `createPortal()` API](https://reactjs.org/docs/portals.html).
```

We don't need to really React in this case?

Author: I think it's helpful here to differentiate between `Portal` (the component) vs. any generic portal generated by the `createPortal()` API.

Reviewer 2: The distinction already seems clear from "Portal is a utility _component_"

Author: > Looking at the Material UI docs, we don't seem to be consistent in the use of backticks for component names. It does seem the be the predominant style, but it isn't necessarily better. The capitalisation seems sufficient, along with an indefinite article: "a portal" vs. "Portal"

I thought we had established a house rule for this in the Style guide, but I'm not seeing anything about it now that I'm looking through it. Must have been deep in some PR comments somewhere. 

In the Base docs, I've been trying to stick to a consistent format of backticks and capitalization anytime I'm referring to the component specifically, in the interest of being as clear and precise as possible.

In any case, this is a question that applies to all of the Base docs, so I propose that we leave it as is for now, and continue this conversation about formatting to my next PR which will be a final pass covering all of these pages.

It gives you the functionality of `createPortal()` in a convenient component form.

:::info
💡 According to [the React docs](https://reactjs.org/docs/portals.html), portals are useful when "you need the child element to visually 'break out' of its container."

> :::

Reviewer 3: Would it read better by grouping concerns? First, the main message, second the side info?

<img width="781" alt="Screenshot 2022-05-04 at 17 38 32" src="https://user-images.githubusercontent.com/3165635/166718348-212aa290-1ba7-447a-b3b0-254d0cb58696.png">

Author: Re: [Marija's comment here](https://github.com/mui/material-ui/pull/32157#discussion_r864975703) I'm thinking about moving this info message down into the "Basic usage" section and elaborating on it a bit.

> The `Portal` component is used internally by Material UI's [`Modal`](material-ui/react-modal/) and [`Popper`](material-ui/react-popper/) components.

Reviewer 3: At this point it's:

- The URL is broken 404
- MUI in general?
- Keep the user in MUI Base?

```suggestion
The `Portal` component is used internally by MUI's [`Modal`](/base/react-modal/) and [`Popper`](/base/react-popper/) components.
``` @

## Basic usage

Normally, children of a component are rendered within that component's DOM tree.

> But sometimes it's necessary to mount a child to a different location in the DOM.

Reviewer 4: Why? We should give here example/explanation

Author: I did a little bit of rearranging to elaborate on this. Let me know what you think!

> But sometimes it's necessary to mount a child to a different location in the DOM.

Reviewer 2: ```suggestion
But sometimes it's necessary to mount a child at a different location in the DOM.
```

{{"demo": "SimplePortal.js"}}
## Basic usage

Normally, children of a component are rendered within that component's DOM tree.
But sometimes it's necessary to mount a child to a different location in the DOM.

The `Portal` component accepts a `container` prop that passes a `ref` to the DOM node where its children will be mounted.

> ## Server-side

Reviewer 3: I think that we could keep a header for this and explain in more detail what we have in https://mui.com/material-ui/react-portal/#server-side. It makes it easier to skim the content

Maybe something like this?

<img width="770" alt="Screenshot 2022-05-04 at 17 35 01" src="https://user-images.githubusercontent.com/3165635/166717548-1430c37d-f681-48f6-93fd-9e6d5376d86f.png">

> ⚠️ **Note:** React [does not support](https://github.com/facebook/react/issues/13097) the [`createPortal()` API on the server](https://reactjs.org/docs/portals.html)—client-side hydration is necessary to render the child elements.

Reviewer 2: ```suggestion
⚠️ **Note:** React [doesn't support](https://github.com/facebook/react/issues/13097) the [`createPortal()` API on the server](https://reactjs.org/docs/portals.html)—client-side hydration is necessary to render the child elements.
```

# Portal

> <p class="description">The Portal component lets you render its children into a DOM node that exists outside of its own DOM hierarchy.</p>

Reviewer 2: For most other components we describe the general use, so:
 ```suggestion
<p class="description">A portal lets you render its children into a DOM node that exists outside of its own DOM hierarchy.</p>
```

Author: I've been doing a mix of both across the Base docs—sometimes one makes more sense than the other. See: [Button](https://mui.com/base/react-button/) vs [Click-away listener](https://mui.com/base/react-click-away-listener/).

In the case of the Portal, I think it makes sense either way.

The `Portal` component is used internally by the [`ModalUnstyled`](/base/react-modal/) and [`Popper`](/base/react-popper/) components.

## Basic usage

Normally, children of a component are rendered within that component's DOM tree.
But sometimes it's necessary to mount a child at a different location in the DOM.

:::info
💡 According to [the React docs](https://reactjs.org/docs/portals.html), portals are useful when "you need the child element to visually 'break out' of its container"—for instance, modals and tooltips, which need to exist outside of the normal flow of the document.
:::

The `Portal` component accepts a `container` prop that passes a `ref` to the DOM node where its children will be mounted.

The following demo shows how a `<span>` nested within a `Portal` can be appended to a node outside of its own DOM hierarchy—click **Mount children** to see how it behaves:

{{"demo": "SimplePortal.js"}}

## Server-side

:::warning
⚠️ **Note:** React doesn't support the [`createPortal()` API](https://reactjs.org/docs/portals.html) on the server.
See [this GitHub issue](https://github.com/facebook/react/issues/13097) for details.
:::

> The `Portal` component cannot be used to render child elements on the server—client-side hydration is necessary.

Reviewer 2: ```suggestion
Portal can't be used to render child elements on the server—client-side hydration is necessary.
```

Author: Same as [this comment](https://github.com/mui/material-ui/pull/32157#discussion_r868209990), I propose we return to this question in the next PR that I will open for a final pass across all Base docs at once.
